### ISSUE TITLE
Communication problems with OTP 24

### ISSUE THREAD

**Issue by dtip** [Pyrlang/Pyrlang#80]
Hi folks. I'm seeing 2 comms problems with Erlang/OTP 24 with the latest Pyrlang and Pyrlang/Term installed with `pip` from the GitHub `master` branch:

1. `net_adm:ping()` hangs forever. It looks like the connection with the pyrlang node works but there's no reply received.
2. `gen_server:call` times out. The call seems to be received and handled correctly by the pyrlang node, and a reply is sent, but the reply isn't picked up as a reply to the call.

## GenServer.call example

Using Elixir 1.12:
```
iex(dev@127.0.0.1)4> GenServer.call(server, :ping)
** (exit) exited in: GenServer.call({:server, :"pyrlang@127.0.0.1"}, :ping, 5000)
    ** (EXIT) time out
    (elixir 1.12.0) lib/gen_server.ex:1024: GenServer.call/3
iex(dev@127.0.0.1)4> flush()
{{[:alias], #Reference<0.2618808245.3977314309.6193>}, :pong}
:ok
```

When I tried the same with Erlang 24, the call times out still but the `flush().` is empty. Not sure why that would be.

**Comment by dtip** [comment]
@s2hc-johan any luck replicating?

**Comment by s2hc-johan** [comment]
Yea I managed to. Trying to figure out what has changed. it looks like the communication between nodes has changed. I'm seeing sometihg with an atom called `alias` that I don't understand.

But I'm on it

**Comment by dtip** [comment]
Thanks, @s2hc-johan !

**Comment by jozip** [comment]
There are a few things in [OTP 24](https://blog.erlang.org/My-OTP-24-Highlights/) that appear to trip things up, most notably the concept of [process aliases](https://erlang.org/doc/reference_manual/processes.html#process-aliases). [EEP 53](https://www.erlang.org/erlang-enhancement-proposals/eep-0053.html) does however state that "The alias feature is a pure extension, so there are no real backwards compatibility issues", so I guess it's mostly an issue of Pyrlang being confused by the additional stuff being sent.

**Comment by s2hc-johan** [comment]
@jozip ok, don’t know how that interface is, and if it has changed. From my quick look it seems to be interfering with the ref, and it’s actually otp24 that can’t handle the return pyrlang sends.

I’m guessing it’s this code that fails due to the pattern looking different in otp24.
https://github.com/Pyrlang/Pyrlang/blob/d6b90df4534e7ddf941008a1fac634b20316a496/pyrlang/gen/decorators.py#L94-L107

**Comment by s2hc-johan** [comment]
looking at the code of otp, it looks like they’re using an improper list for this,

https://github.com/erlang/otp/blob/8b21f5c8aaa3e13046c26ecb889dae4f5dcdd70c/lib/stdlib/src/gen.erl#L281-L284

Maybe there is a roundtrip bug there so it doesn’t return as it should. I’ll check some more tomorrow

**Comment by jozip** [comment]
@s2hc-johan yeah, that doesn't look right. Surely it's a mistake?

Anyhow, running `make example10a` and trying `gen_server:call({my_process, 'py@127.0.0.1'}, <<"hello">>)` from OTP 24 yields
`got inbox <__main__.MyProcess object at 0x7f4a7f0eca20>, <pyrlang.match.Match object at 0x7f4a7f0ece48>, (Atom('$gen_call'), (<1622989162.177.0 @ erl24@127.0.0.1>, ([Atom('alias')], Ref<1622989162,53450,2709061633,4128230993>@erl24@127.0.0.1)), b'hello')`

The same call from OTP 23.3.4.2 looks like this:
`got inbox <__main__.MyProcess object at 0x7f4a7f0eca20>, <pyrlang.match.Match object at 0x7f4a7f0ece48>, (Atom('$gen_call'), (<1622844056.98.0 @ erl@127.0.0.1>, Ref<1622844056,214818,2977431554,1039548637>@erl@127.0.0.1), b'hello')`

The `Ref` becoming `([Atom('alias')], Ref)` looks like trouble.

**Comment by s2hc-johan** [comment]
@jozip yea I've just verified that a roundtrip `binary -> term -> binary` with an improper list isn't consistent. so we start with an improper list and end up with a tuple of the elements and the tail instead. I'll start looking at a solution and my guess is that this will be fixed by a pyrlang-term release. I've created an issue in pyrlang-term to track it there.

https://github.com/Pyrlang/Term/issues/23
